hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/ReliableTopicProxy.java

```diff
@@ -255,4 +277,139 @@ protected void postDestroy() {
+            throw (RuntimeException) peel(e, null,
+                    String.format("Failed to publish messages: %s on topic: %s", payload, getName()));
+        }
+    }
+
+    @Override
+    public CompletionStage<Void> publishAllAsync(@Nonnull Collection<? extends E> payload) {
+        checkNotNull(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+        checkNoNullInside(payload, NULL_MESSAGE_IS_NOT_ALLOWED);
+
+        // RU_COMPAT_4_0
+        checkClusterVersion(Versions.V4_1);
+
+        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();
+        try {
+            List<ReliableTopicMessage> messages = payload.stream()
+                    .map(m -> new ReliableTopicMessage(nodeEngine.toData(m), thisAddress))
+                    .collect(Collectors.toList());
+            switch (overloadPolicy) {
+                case ERROR:
+                    addAsyncOrFail(payload, returnFuture, messages);
+                    break;
+                case DISCARD_OLDEST:
+                    addAsync(messages, OverflowPolicy.OVERWRITE);
+                    break;
+                case DISCARD_NEWEST:
+                    addAsync(messages, OverflowPolicy.FAIL);
+                    break;
+                case BLOCK:
+                    addAsyncAndBlock(payload, returnFuture, messages, INITIAL_BACKOFF_MS);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
+            }
+        } catch (Exception e) {
+            throw (RuntimeException) peel(e, null,
+                    String.format("Failed to publish messages: %s on topic: %s", payload, getName()));
+        }
+
+        return returnFuture;
+    }
+
+    private void addAsyncOrFail(@Nonnull Collection<? extends E> payload, InternalCompletableFuture<Void> returnFuture,
+                                List<ReliableTopicMessage> messages) {
+        ringbuffer.addAllAsync(messages, OverflowPolicy.FAIL).whenCompleteAsync((id, t) -> {
+            if (t != null) {
+                returnFuture.completeExceptionally(t);
+            } else if (id == -1) {
+                returnFuture.completeExceptionally(new TopicOverloadException(
+                        "Failed to publish messages: " + payload + " on topic:" + getName()));
+            } else {
+                returnFuture.complete(null);
+                messages.forEach(p -> localTopicStats.incrementPublishes());
+            }
+        });
+    }
+
+    private InternalCompletableFuture<Void> addAsync(List<ReliableTopicMessage> messages, OverflowPolicy overflowPolicy) {
+        InternalCompletableFuture<Void> returnFuture = new InternalCompletableFuture<>();
+        ringbuffer.addAllAsync(messages, overflowPolicy).whenCompleteAsync((id, t) -> {
+            if (t != null) {
+                returnFuture.completeExceptionally(t);
+            }
+            returnFuture.complete(null);
```

Comment:
Move this to the else block so we don't complete or increment when there's an exception. Same on client side.

Reply:
Done